lessons/projects/snake/index.md

```diff
@@ -0,0 +1,149 @@
+
+0. Promysli si, jak bude hra fungovat a jak přeneseme mřížku s hadem
+z příkazové řádky do grafického okna.
+1. Vykresli hada do grafického okna (ve formě barevných čtverců)
+2. Přidej funkci, která bude s hadem hýbat.
+3. Umožni změnit směr hada pomocí klávesnice.
+4. Nenech hada utéct z herní plochy a nabourat do sebe sama.
+5. Přidej hadovi jídlo a zajisti, aby po jídle rostl.
+6. Vyměň barevné čtverečky za opravdovou grafiku.
+
+Po těchto krocích budeš mít základní hru, ale tou to nekončí, právě napak!
+Máš vlastní hru jějímuž fungování rozumíš jako nikdo jiný a to je to pravé pro
+přidávání dalších možností. Fantazii se meze nekladou. Například:
+
+1. Ve hře mohou být dva nebo třeba tři hadi najednou - každý ovládaný
+jinými klávesami navzájem soupeřící o jídlo.
+2. Kromě jídla se mohou na ploše objevovat i jiné objekty - překážky,
+do kterých nesmí had narazit, otrávené jídlo, které hada zkrátí atp.
+3. Hrací plocha může být nekonečná a když z ní had vyleze, objeví se
+na druhé straně.
+
+## Z příkazové řádky do grafické aplikace
+
+V příkazové řádce měl had souřadnice označující řádek a sloupec. V grafické
+aplikaci to bude podobné, ale protože pixelů na obrazovce je mnohem více, budeme
+si muset vytvořit pomyslnou síť stejně velkých čtverců, které nám nahradí
+řádky a sloupce. Velikost takového čtverce bude konstanta, kterou se vyplatí
+mít po celou dobu hry k dispozici, aby se podle ní daly vypočítat
+souřadnice k vykreslení obrázků. Pro začátek řekněme, že ideální velikost
+takového čtverce bude 64×64 pixelů.
+
+Z velikosti takového čtverce, kterou si můžeme v budoucnu libovolně změnit,
+a velikosti okna aplikace můžeme vypočítat, kolik se nám do okna takových
+čtverců vejde na šířku a na výšku a tím i zjistit, kolik pomyslných
+sloupců a řádků bude naše hrací plocha mít.
+
+## Vykreslení hada
+
+Abychom mohli hada vykreslit, potřebujeme si pro začátek uložit jeho souřadnice.
+K tomu můžeš použít seznam dvojic - stejně jako v domácím projektu. Pokud chceš,
+můžeš si pro hada vytvořit i vlastní třídu, kde bude seznam souřadnic jako jeden
+z atributů.
+
+> [note]
+> Záleží jen na tobě, jak moc budeš chtít hada rozšiřovat. I když se totiž může
+> na začátku zdát vlastní třída jako zbytečná komplikace, později může přispět
+> k přehlednějšímu kódu a snazšímu rozšíření třeba o další hady.
+
+{{ figure(
+    img=static('coords.svg'),
+    alt="Had na „šachovnici“ se souřadnicemi",
+) }}
+
+Když už je had definován, budeme potřebovat jednoduchou funkci, která
+na ta správná místa umístí obrázky. Pro začátek si vystačíme se zeleným
+čtvercem. Obrázek si [stáhni zde]({{ static('green.png')}} ) a ulož
+do složky k programu.
+
+Na lekci jsme načtený obrázek ukládali do tzv. Spritů, abychom jimi
+pak mohli pohybovat a dosáhli plynulého pohybu. Tady bude situace jednodušší,
+protože se obejdeme bez plunylého pohybu a tak nám stačí obrázek několikrát
+vykreslit na vyprázdněnou plochu.
```

Comment:
Důvod pro `Sprite` proti `blit` je spíš že se dá otáčet (a pak obarvovat atd.)

Reply:
Má poznámka o plynulém pohybu vycházela z toho, že `Sprite` se dá vložit do `batch` a tím vykreslovat rychleji a v dynamických hrách s plynulým pohybem mnoha objektů tak zajistit dostatečnou rychlost a plynulost. Mělo to fungovat jako takový odkaz na Asteroidy, kde se tohle řeší. Něco jako: „V Asteroidech kolegyně vedle tebe řeší plynulý pohyb, ale ty se bez toho obejdeš“ pro ty méně zdatné a bojácné a „V Asteroidech na tebe čeká nová výzva s podobě plynulého pohybu mnoha objektů najednou“ pro ty odvážné.

Má smysl tu moji myšlenku rozvinout a v tomhle duchu to upravit, nebo to naopak zjednodušit a jako důvod nevyužití `Sprite` dát že je nepotřebujeme otáčet?

Reply:
Myslím že má cenu používat `Sprite`.

Jo, ještě jedna věc: že Sprite řeší průhlednost. Když používáš jen `blit`, je potřeba docela neintuitivním „zaříkadlem“ zapnout alpha blending:

```python
    pyglet.gl.glEnable(pyglet.gl.GL_BLEND)
    pyglet.gl.glBlendFunc(pyglet.gl.GL_SRC_ALPHA, pyglet.gl.GL_ONE_MINUS_SRC_ALPHA)
```

Na workshopu to jepro ušetření času.

Reply:
Jaký zvolit kompromis mezi optimalizací a složitostí implementace?

* Jednoduše vytvořit `Sprite`, vykreslit, zahodit
* Udržovat si seznamy `Sprite` (pro každý kousek hada jeden, pro každé ovoce jeden), měnit jejich souřadnice a při použití obrázků i obsah a po jednom je vykreslovat
* Udržovat si jejich seznam jako v předchozím bodě a navíc mít všechny v `batch` a vykreslovat je tedy najednou

První varianta sice není optimální, ale zase je to jednoduché.

Reply:
Ostatní můžeš nechat na rozšíření :)